builtin/v10/init/invariants.go

```diff
@@ -0,0 +1,80 @@
+package init
+
+import (
+	addr "github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/manifest"
+	"github.com/ipfs/go-cid"
+	cbg "github.com/whyrusleeping/cbor-gen"
+
+	"github.com/filecoin-project/go-state-types/builtin"
+	"github.com/filecoin-project/go-state-types/builtin/v10/util/adt"
+)
+
+type StateSummary struct {
+	AddrIDs map[addr.Address]abi.ActorID
+	NextID  abi.ActorID
+}
+
+// Checks internal invariants of init state.
+func CheckStateInvariants(st *State, tree *builtin.ActorTree, actorCodes map[string]cid.Cid) (*StateSummary, *builtin.MessageAccumulator) {
+	acc := &builtin.MessageAccumulator{}
+	store := tree.Store
+
+	acc.Require(len(st.NetworkName) > 0, "network name is empty")
+	acc.Require(st.NextID >= builtin.FirstNonSingletonActorId, "next id %d is too low", st.NextID)
+
+	initSummary := &StateSummary{
+		AddrIDs: nil,
+		NextID:  st.NextID,
+	}
+
+	lut, err := adt.AsMap(store, st.AddressMap, builtin.DefaultHamtBitwidth)
+	if err != nil {
+		acc.Addf("error loading address map: %v", err)
+		// Stop here, it's hard to make other useful checks.
+		return initSummary, acc
+	}
+
+	initSummary.AddrIDs = map[addr.Address]abi.ActorID{}
+	reverse := map[abi.ActorID]addr.Address{}
+	var value cbg.CborInt
+	err = lut.ForEach(&value, func(key string) error {
+		actorId := abi.ActorID(value)
+		keyAddr, err := addr.NewFromBytes([]byte(key))
+		if err != nil {
+			return err
+		}
+
+		acc.Require(keyAddr.Protocol() != addr.ID, "key %v is an ID address", keyAddr)
+		acc.Require(keyAddr.Protocol() <= addr.Delegated, "unknown address protocol for key %v", keyAddr)
+		acc.Require(actorId >= builtin.FirstNonSingletonActorId, "unexpected singleton ID value %v", actorId)
+
+		foundAddr, found := reverse[actorId]
+		isPair := (keyAddr.Protocol() == addr.Actor && foundAddr.Protocol() == addr.Delegated) ||
+			(keyAddr.Protocol() == addr.Delegated && foundAddr.Protocol() == addr.Actor)
+		dup := found && !isPair
+		acc.Require(!dup, "duplicate mapping to ID %v: %v, %v", actorId, keyAddr, foundAddr)
+		reverse[actorId] = keyAddr
+
+		initSummary.AddrIDs[keyAddr] = actorId
+
+		idaddr, err := addr.NewIDAddress(uint64(actorId))
+		acc.RequireNoError(err, "unable to convert actorId %v to id address", actorId)
+		actor, found, err := tree.GetActorV5(idaddr)
+		acc.RequireNoError(err, "unable to retrieve actor with idaddr %v", idaddr)
+		acc.Require(found, "actor not found idaddr %v", idaddr)
+
+		// we expect the address field to be populated for the below actors
+		if (actor.Code == actorCodes[manifest.EthAccountKey] ||
```

Comment:
I'd like to make two more assertions here:

- If keyAddr is a delegated address, the actor code MUST be one of these 3 types (might be useful having some `isDelegatable` helper, please come up with a better name than I can)
- If an actor in the state tree has an `Address` set, we find a corresponding mapping for it in the init actor (this will need an iteration over the entire actor tree here, or can be a preprocessing step that we do in the top-level iteration before calling the init actor's invariant check).

Reply:
Is #1 dong to be true in the future? From what I understand, the plan is to have delegated addresses for all actors

For #2, this can be added as a cross check between a list built from the top level and the init actor?

Reply:
#1, no it won't, but it will be true for all v10 actors (which is what these invariants are for)
#2 yup, that would work

Reply:
done